src/sleep.rs

```diff
@@ -0,0 +1,376 @@
+        )
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+#[derive(Debug)]
+pub enum RtcWakeLevel {
+    AllLow,
+    AnyHigh,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl RtcWakeLevel {
+    pub fn mode(&self) -> u32 {
+        match self {
+            RtcWakeLevel::AllLow => esp_sleep_ext1_wakeup_mode_t_ESP_EXT1_WAKEUP_ALL_LOW,
+            RtcWakeLevel::AnyHigh => esp_sleep_ext1_wakeup_mode_t_ESP_EXT1_WAKEUP_ANY_HIGH,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub struct RtcWakeupPin {
+    pub pin: AnyInputPin, // this should use RtcInput somehow
+    pub pullup: bool,
+    pub pulldown: bool,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl RtcWakeupPin {
+    pub fn new(pin: AnyInputPin, pullup: bool, pulldown: bool) -> Self {
+        Self {
+            pin,
+            pullup,
+            pulldown,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl fmt::Debug for RtcWakeupPin {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "RtcWakeupPin {{ pin: {:?} pullup {:?} pulldown {:?}}}",
+            self.pin.pin(),
+            self.pullup,
+            self.pulldown
+        )
+    }
+}
+
+/// Will wake up the CPU based on changes in GPIO pins. Is only available for light sleep.
+/// It takes PinDriver as input, which means that any required configurations on the pin
+/// can be done before adding it to the GpioWakeup struct.
+#[derive(Debug)]
+pub struct GpioWakeup<'a> {
+    pub pins: &'a [GpioWakeupPin<'a>],
+}
+
+impl<'a> GpioWakeup<'a> {
+    pub fn new(pins: &'a [GpioWakeupPin<'a>]) -> Self {
+        Self { pins }
+    }
+}
+
+impl<'a> GpioWakeup<'a> {
+    pub fn apply(&self) -> Result<(), EspError> {
+        for pin in self.pins.iter() {
+            let intr_level = match pin.wake_level {
+                Level::Low => gpio_int_type_t_GPIO_INTR_LOW_LEVEL,
+                Level::High => gpio_int_type_t_GPIO_INTR_HIGH_LEVEL,
+            };
+            esp!(unsafe { gpio_wakeup_enable(pin.pin.pin(), intr_level) })?;
+        }
+        esp!(unsafe { esp_sleep_enable_gpio_wakeup() })?;
+        Ok(())
+    }
+}
+
+pub struct GpioWakeupPin<'a> {
+    pub pin: PinDriver<'a, AnyInputPin, Input>,
+    pub wake_level: Level,
+}
+
+impl<'a> GpioWakeupPin<'a> {
+    pub fn new(
+        pin: PinDriver<'a, AnyInputPin, Input>,
+        wake_level: Level,
+    ) -> Result<Self, EspError> {
+        Ok(Self { pin, wake_level })
+    }
+}
+
+impl<'a> fmt::Debug for GpioWakeupPin<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "GpioWakeupPin {{ pin: {:?} wake_level: {:?} }}",
+            self.pin.pin(),
+            self.wake_level
+        )
+    }
+}
+
+/// Will wake up the CPU when the number of positive edges higher than a threshold
+/// is detected on the UART RX pin. Is only available for light sleep.
+pub struct UartWakeup<'a> {
+    pub uart: &'a UartDriver<'a>,
```

Comment:
Why `UartDriver` and not the actual UART peripheral? The deeper question is... should the ESP IDF UART driver be installed and working (this is `UartDriver`) for the wakeup to happen or not? Depending on the answer to that we need to pass here either the UART driver, or the UART peripheral.